codec/src/codec.rs

```diff
@@ -3,53 +3,122 @@
+pub trait Write {
+    /// Encodes this value by writing to a buffer.
+    ///
+    /// Implementations should panic if the buffer doesn't have enough capacity.
     fn write(&self, buf: &mut impl BufMut);
+}
 
+/// Trait for types that can be read/decoded from a buffer.
+///
+/// The `Cfg` type parameter allows for configuration during the read process. For example, it can
+/// be used to limit the maximum size of allocated buffers for safety when decoding untrusted data.
+/// Use `()` for types that do not require configuration.
+pub trait Read<Cfg = ()>: Sized {
+    /// Reads a value from the buffer using the provided configuration `cfg`, consuming the
+    /// necessary bytes.
+    ///
+    /// Returns an error if decoding fails (e.g., invalid data, not enough bytes initially).
+    fn read_cfg(buf: &mut impl Buf, cfg: Cfg) -> Result<Self, Error>;
+}
+
+/// Trait for types that can be encoded to a buffer.
+pub trait Encode: Write {
     /// Returns the encoded length of this value.
+    ///
+    /// This method MUST return the exact number of bytes that will be written by `write()`.
     fn len_encoded(&self) -> usize;
 
-    /// Encodes a value to bytes.
+    /// Encodes a value to a `BytesMut` buffer.
+    ///
+    /// Panics if the `write` implementation does not write the expected number of bytes.
+    ///
+    /// (Provided method).
     fn encode(&self) -> BytesMut {
         let len = self.len_encoded();
         let mut buffer = BytesMut::with_capacity(len);
         self.write(&mut buffer);
-        assert_eq!(buffer.len(), len);
+        assert_eq!(buffer.len(), len, "write() did not write expected bytes");
         buffer
     }
+}
 
-    /// Reads a value from a buffer, returning an error if there is an error while reading.
-    fn read(buf: &mut impl Buf) -> Result<Self, Error>;
+// Automatically implement `Encode` for types with a known size.
+// Otherwise, the type must define its own `len_encoded()` method.
+impl<T: EncodeFixed> Encode for T {
+    fn len_encoded(&self) -> usize {
+        Self::LEN_ENCODED
+    }
+}
 
-    /// Decodes a value from a buffer.
+/// Trait for types that can be decoded from a buffer, ensuring the entire buffer is consumed.
+pub trait Decode<Cfg = ()>: Read<Cfg> + Size {
+    /// Decodes a value from a buffer, ensuring the buffer is fully consumed.
+    ///
+    /// For types with a known size, this method first checks that the buffer has the expected size.
     ///
-    /// Returns an error if there is an error while decoding or if there is extra data remaining
-    /// after decoding the value from the buffer.
-    fn decode<B: Buf>(mut buf: B) -> Result<Self, Error> {
-        let result = Self::read(&mut buf);
+    /// (Provided method).
+    fn decode_cfg(mut buf: impl Buf, cfg: Cfg) -> Result<Self, Error> {
+        // If we can, before reading, check that the buffer has the expected size.
+        if let Some(size) = Self::FIXED_SIZE {
```

Comment:
If we dropped `FIXED_SIZE` (to only have LEN_ENCODED), we would just lose this early optimizaation?